sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/dao/room/EntityTypeConverters.kt

```diff
@@ -177,6 +201,77 @@ class EntityTypeConverters {
     }
 }
 
+/**
+ * A client writable copy is considered a new object with only the properties set on the object
+ * that are writable on bridge.  All other fields sent to bridge will be ignored anyways.
+ * @return a ScheduledActivity that can be sent to bridge
+ */
+fun ScheduledActivityEntity.clientWritableCopy(): ScheduledActivity {
+    val schedule = ScheduledActivity()
+    schedule.guid = guid
+    schedule.startedOn = startedOn?.let { org.joda.time.DateTime(it.toEpochMilli()) }
+    schedule.finishedOn = finishedOn?.let { org.joda.time.DateTime(it.toEpochMilli()) }
+    schedule.clientData = clientData?.data
+    return schedule
+}
+
+/**
+ * A bridge metadata copy is considered a new object with only the properties set on the object
+ * that the ArchiveUtil.createMetaDataFile() function is concerned with.
+ * @return a ScheduledActivity that can be pass to the BridgeDataProvider and have the metadata JSON included.
+ */
+fun ScheduledActivityEntity.bridgeMetadataCopy(): ScheduledActivity {
+    var schedule = ScheduledActivity()
+    val immutableFieldMap = HashMap<String, String>()
+
+    scheduledOn?.let {
+        // TODO: mdephillips 10/14/18 better way to convert from 3tenbp LocalDateTime to jodatime DateTime?
+        val dateTime = DateTime.now()
+                .withYear(it.year)
+                .withDayOfYear(it.dayOfYear)
+                .withHourOfDay(it.hour)
+                .withMinuteOfHour(it.minute)
+                .withSecondOfMinute(it.second)
+        val dateTimeStr = ISODateTimeFormat.dateTime().withOffsetParsed().print(dateTime)
+        immutableFieldMap.put("scheduledOn", dateTimeStr)
+    }
+
+    schedulePlanGuid?.let {
+        immutableFieldMap.put("schedulePlanGuid", it)
+    }
+
+    if (immutableFieldMap.isNotEmpty()) {
+        // TODO: mdephillips 10/14/18 sync with josh or dwayne to make scheduledOn and schedulePlanGuid mutable
+        // TODO: mdephillips 10/14/18 so I don't have to do this json non-sense
+        var scheduleJson = "{"
+        immutableFieldMap.keys.forEachIndexed { i, s ->
+            scheduleJson = "$scheduleJson \"$s\" : \"${immutableFieldMap[s]}\""
+            if (i < (immutableFieldMap.keys.size - 1)) {
+                scheduleJson = "$scheduleJson,"
+            }
+        }
+        scheduleJson = "$scheduleJson}"
+        try {
+            schedule = RestUtils.GSON.fromJson(scheduleJson, ScheduledActivity::class.java)
+        } catch (e: JsonSyntaxException) {
+            e.printStackTrace()
```

Comment:
Log?

Reply:
I wasn't sure how to log from an extension function, because it doesn't really have a place to put the logger object.  Maybe I can just use LogExt.e() ?

Reply:
Ah. We could also make a static Logger on ScheduledActivityEntity or just do LoggerFactory.getLogger(ScheduledActivityEntity.class).warn(msg) inline